### PR TITLE
PlatformView to set the correct overlay view frame, merge thread at the endFrame

### PR DESCRIPTION
**Pull request by cyanglaz**
1. This PR makes the OverlayView(which is masked by OverlayWrapper) to be at the correct frame. The new frame of the OverlayView is the same as the FlutterView. Because the Overlay is at the correct frame, we don't need to add a transform matrix to the skia canvas, so that skia doesn't draw off canvas.

This somehow fixes the flickering issue in https://github.com/flutter/flutter/issues/103076, I'm not sure the exact reason behind it but it seems like it is related to Skia trying to draw off canvas.

2. Makes the thread merging happen at the end of the last frame on raster thread. This prevents a new frame starts on the raster thread before the thread merging is completed. 

This fixes some occasional unsynchronized frame, also observed in https://github.com/flutter/flutter/issues/103076

I have observed another thread synchronization issue during thread un-merging. It seems a little more complicated so I'll create a separate PR for it .

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by zanderso** [comment]
Use the template here https://github.com/flutter/flutter/issues/new?assignees=&labels=cp%3A+review&template=7_cherry_pick.yml&title=%5BCP%5D+%3Ctitle%3E

**Comment by zanderso** [comment]
This will be easier to approve for a cherry-pick if it can have a test.

**Comment by cyanglaz** [comment]
> This will be easier to approve for a cherry-pick if it can have a test.

I'm adding tests now. Building the test locally took some time and I just put the PR up before it :)

**Comment by cyanglaz** [comment]
@iskakaushik Do you mind take a look at code for both update?
@blasten Do you mind take a look at the code, specifically related to unobstructed platform views?

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm [hunk 1]
```diff
@@ -609,14 +616,18 @@
...
-  // Set a unique view identifier, so the overlay wrapper can be identified in unit tests.
+  // Set a unique view identifier, so the overlay_view_wrapper can be identified in XCUITests.
   overlay_view_wrapper.accessibilityIdentifier =
       [NSString stringWithFormat:@"platform_view[%lld].overlay[%lld]", view_id, overlay_id];
 
   UIView* overlay_view = layer->overlay_view.get();
   // Set the size of the overlay view.
   // This size is equal to the device screen size.
-  overlay_view.frame = flutter_view_.get().bounds;
+  overlay_view.frame = [flutter_view_.get() convertRect:flutter_view_.get().bounds
```
Comment (blasten) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm:626]: have you seen any side effects when the frame size is changed in every frame?  Try running https://github.com/flutter/flutter/tree/195a1cc413bde5f7c1d194204608c4bc20659124/dev/benchmarks/platform_views_layout with this change. That test has an animated container that should cause this value to change after each frame https://github.com/flutter/flutter/blob/195a1cc413bde5f7c1d194204608c4bc20659124/dev/benchmarks/platform_views_layout/lib/main.dart#L121

**Comment by cyanglaz** [comment]
@zanderso Looks like the original issue https://github.com/gskinnerTeam/flutter-wonders-app/issues/98 was magically fixed. Since it is not a regression and requires a specific device/steps to reproduce, I would suggest to not cherry-pick in. WDTY?

**Comment by zanderso** [comment]
I'm not sure that I trust that it was magically fixed. Is there a version of the app where it repros locally for you without your fix?

**Comment by cyanglaz** [comment]
> I'm not sure that I trust that it was magically fixed. Is there a version of the app where it repros locally for you without your fix?

I wasn't able to repro the same exact issue they were experiencing. Not with local build, flutter stable/master, nor the TestFlight build. 

Mine looks a little bit different as shown in https://github.com/flutter/flutter/issues/103076 and it repros even in stable. Their video capture, however, does look like some sort of syncing issue between PlatformView and Overlay. Since the fix on the thread merging in this PR fixed the flickering issue that I was experiencing, I guess the same code might fix the issue they saw.

I asked them to test their app with this PR and see if it fixes the glitch that they saw. but it seem they couldn't reproduce the issue anymore.

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm (hunk 1 above)
Comment (cyanglaz) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm:626]: Good point. I ran the benchmarks, there doesn't seem to be significant difference:

With my change:
```
  "average_frame_build_time_millis": 0.2156956521739131,
  "90th_percentile_frame_build_time_millis": 0.27,
  "99th_percentile_frame_build_time_millis": 0.316,
  "worst_frame_build_time_millis": 0.326,
  "missed_frame_build_budget_count": 0,
  "average_frame_rasterizer_time_millis": 4.833215053763439,
  "90th_percentile_frame_rasterizer_time_millis": 6.668,
  "99th_percentile_frame_rasterizer_time_millis": 7.84,
  "worst_frame_rasterizer_time_millis": 9.612,
  "missed_frame_rasterizer_budget_count": 0,
```

master:
```
  "average_frame_build_time_millis": 0.21664516129032269,
  "90th_percentile_frame_build_time_millis": 0.286,
  "99th_percentile_frame_build_time_millis": 0.305,
  "worst_frame_build_time_millis": 0.357,
  "missed_frame_build_budget_count": 0,
  "average_frame_rasterizer_time_millis": 4.918989130434783,
  "90th_percentile_frame_rasterizer_time_millis": 6.905,
  "99th_percentile_frame_rasterizer_time_millis": 8.33,
  "worst_frame_rasterizer_time_millis": 9.434,
  "missed_frame_rasterizer_budget_count": 0,
```

**Comment by zanderso** [comment]
Got it. Not sure I completely follow the timeline, but if the issue was possibly introduced by the 120Hz change, then we should queue this up for a cherry-pick. Otherwise, that is if it's just a fix for a pre-existing issue, then I don't think we need a cherry-pick.

**Comment by iskakaushik** [comment]
This wasn't introduced by the 120hz work, rendering at high frame rates only made a previously present issue more apparent. I was able to reproduce this with the 120hz change reverted as well.

**Comment by cyanglaz** [comment]
> Got it. Not sure I completely follow the timeline, but if the issue was possibly introduced by the 120Hz change, then we should queue this up for a cherry-pick. Otherwise, that is if it's just a fix for a pre-existing issue, then I don't think we need a cherry-pick.

Yes, the issue was not related to 120Hz. The issue happens randomly, it is a little more reproducible with 120hz enabled just because we draw more frames with 120Hz. 
And sounds good, let's not cherry-pick this. I'm going to lower the priority of the issue.

**Comment by cyanglaz** [comment]
Also a note that this only happens on iPhone 13pro. I'm not sure why, but It is possible that Skia drawing outside of the canvas worked in previous iPhones but not the iPhone 13pro.

**Comment by zanderso** [comment]
Chatted with folks offline. It sounds like a cherry-pick isn't needed. Is this ready to land, though?

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm (hunk 1 above)
Comment (blasten) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm:626]: did you see any issue in general? if you run the app, and play with it?  e.g. Could this change introduce jank or glitches?

Comment (cyanglaz) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm:626]: no, I didn't see this change introduce new janks or glitches.

**Comment by cyanglaz** [comment]
friendly ping @iskakaushik for review

File: shell/platform/darwin/ios/ios_external_view_embedder.mm [hunk 2]
```diff
@@ -84,7 +84,7 @@
 void IOSExternalViewEmbedder::EndFrame(bool should_resubmit_frame,
                                        fml::RefPtr<fml::RasterThreadMerger> raster_thread_merger) {
   TRACE_EVENT0("flutter", "IOSExternalViewEmbedder::EndFrame");
-  FML_CHECK(platform_views_controller_);
```
Comment (iskakaushik) [inline shell/platform/darwin/ios/ios_external_view_embedder.mm:87]: why is this deleted?

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm [hunk 3]
```diff
@@ -277,7 +277,6 @@
     // Eventually, the frame is submitted once this method returns `kSuccess`.
     // At that point, the raster tasks are handled on the platform thread.
     CancelFrame();
-    raster_thread_merger->MergeWithLease(kDefaultMergedLeaseDuration);
```
Comment (iskakaushik) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm:280]: why is this removed?

File: shell/platform/darwin/ios/ios_external_view_embedder.mm (hunk 2 above)
Comment (cyanglaz) [reply shell/platform/darwin/ios/ios_external_view_embedder.mm:87]: This is a CHECK, so it is essentially crashing the app if `platform_views_controller_` is null, which is exactly what the next will do. 

I think this check shouldn't be here at all.

Comment (cyanglaz) [reply shell/platform/darwin/ios/ios_external_view_embedder.mm:87]: I don't think there's any reason it would happen in the current code, but it might happen if rasterizer's logic changes.

Good point to have a failing message, I will add the check back and add a failing message.

Comment (cyanglaz) [reply shell/platform/darwin/ios/ios_external_view_embedder.mm:87]: I thought about it but not sure what additional message to add that would be useful. We assume `platform_views_controller_` is not null here and it crashes if it is null. 

I could add something like "platform_views_controller_ must not be null" but it is redundant as crashing at `latform_views_controller_->EndFrame(should_resubmit_frame, raster_thread_merger);` would self-explained it.

**Review by iskakaushik — APPROVED** [review]
LGTM

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm (hunk 1 above)
Comment (cyanglaz) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm:626]: friendly ping @blasten 
Did you have other suggestions/comments?

**Comment by jmagman** [comment]
@cyanglaz is this ready to merge?

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm (hunk 1 above)
Comment (blasten) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm:626]: no more suggestions. Thanks for looking!

**Review by blasten — APPROVED** [review]
LGTM
